### PR TITLE
RavenDB-12492: Fixed Weird multiplicative factor on MapReduce

### PR DESCRIPTION
**Pull request by redknightlois**

### CODE REVIEW

**Comment by redknightlois** [comment]
@arekpalinski need some review here... I think it is right, but I am not that acquainted with the inner workings of this part of the code.

File: src/Raven.Server/Documents/Indexes/MapReduce/MapReduceResultsStore.cs [hunk 1]
```diff
@@ -52,34 +55,38 @@ public MapReduceResultsStore(ulong reduceKeyHash, MapResultsStorageType type, Tr
...
-            ModifiedPages = new HashSet<long>();
-            FreedPages = new HashSet<long>();
+            if (!_alreadyInitializedTrees.Contains(treeName))
+            {
+                Tree.PageModified += (page, flags) =>
+                {
+                    if ((flags & PageFlags.Overflow) == PageFlags.Overflow)
+                        return;
 
-            _mapReduceContext.PageModifiedInReduceTree += page => FreedPages.Remove(page);
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/MapReduce/MapReduceResultsStore.cs:64]: Why did you remove `PageModifiedInReduceTree`? 
It made that  `RavenDB_11643.PageModificationInAnyTreeMustRemoveItFromListOfFreedPagesInAllStores` failed (http://jenkins.hibernatingrhinos.com/job/PR_Tests_v4.1/2220/)

Please check https://github.com/ravendb/ravendb/pull/7690/ it was added there to handle edge case

File: src/Raven.Server/Documents/Indexes/MapReduce/MapReduceResultsStore.cs [hunk 2]
```diff
@@ -52,34 +55,38 @@ public MapReduceResultsStore(ulong reduceKeyHash, MapResultsStorageType type, Tr
...
+
         private void InitializeTree(bool create)
         {
             var treeName = ReduceTreePrefix + _reduceKeyHash;
             var options = _tx.LowLevelTransaction.Environment.Options.RunningOn32Bits ? TreeFlags.None : TreeFlags.LeafsCompressed;
             Tree = create ? _tx.CreateTree(treeName, flags: options) : _tx.ReadTree(treeName);
 
-            ModifiedPages = new HashSet<long>();
-            FreedPages = new HashSet<long>();
+            if (!_alreadyInitializedTrees.Contains(treeName))
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/MapReduce/MapReduceResultsStore.cs:66]: How this can be called multiple times for the same tree instance? `InitializeTree` is supposed to be called just once

**Review by ayende — APPROVED** [review]

**Comment by ayende** [comment]
This looks good, but note the failing test

File: src/Raven.Server/Documents/Indexes/MapReduce/MapReduceResultsStore.cs (hunk 2 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/Indexes/MapReduce/MapReduceResultsStore.cs:66]: Maybe by IterateTrees?

File: src/Raven.Server/Documents/Indexes/MapReduce/MapReduceResultsStore.cs [hunk 3]
```diff
@@ -52,34 +55,38 @@ public MapReduceResultsStore(ulong reduceKeyHash, MapResultsStorageType type, Tr
             }
         }
 
+        private readonly HashSet<string> _alreadyInitializedTrees = new HashSet<string>();
```
Comment (grisha-kotler) [inline src/Raven.Server/Documents/Indexes/MapReduce/MapReduceResultsStore.cs:58]: StringComparer.OrdinalIgnoreCase?

File: src/Raven.Server/Documents/Indexes/MapReduce/MapReduceResultsStore.cs (hunk 2 above)
Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/MapReduce/MapReduceResultsStore.cs:66]: It's used in debug endpoint only for Map-reduce visualizer purposes, so not very likely I think. Those are different instances of trees anyway so that shoudn't matter

Comment (redknightlois) [reply src/Raven.Server/Documents/Indexes/MapReduce/MapReduceResultsStore.cs:66]: It gets called once per every tree. So in essence it will create a linear list of trees per batch, causing a huge slowdown and lots of wasted effort. Just to give you an idea a 24000 batch will generate 500M calls to `.Remove()`

File: src/Raven.Server/Documents/Indexes/MapReduce/MapReduceResultsStore.cs (hunk 1 above)
Comment (redknightlois) [reply src/Raven.Server/Documents/Indexes/MapReduce/MapReduceResultsStore.cs:64]: Back to the drawing board then. I need to understand what is the rationale of that edge case. Can you elaborate?

Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/MapReduce/MapReduceResultsStore.cs:64]: Please check the commit message, and comment in test of https://github.com/ravendb/ravendb/pull/7690 PR. It would be probably the easiest to debug this test to see what is actually going on.

The main problem was that a page (_p1_) marked as _free_ when processing one reduce tree (_t1_), can be allocated by another reduce tree (_t2_). In the reduce phase, we go over the trees and remove the aggregation results for a pages that were marked as free:

https://github.com/ravendb/ravendb/blob/436d63cd0b9b42473d7dc0945b8f4ebd2be6b2a3/src/Raven.Server/Documents/Indexes/MapReduce/ReduceMapResultsBase.cs#L332-L336

In result, processing _t2_ tree added an entry to PageNumberToReduceResult table for _p1_ page but next processing of _t1_ removed it from that table.

Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/MapReduce/MapReduceResultsStore.cs:64]: So yes, my fix was: if a page is modified (in any tree), let's remove it from _all_ `FreedPages` collections (for every `MapReduceResultsStore`)

File: src/Raven.Server/Documents/Indexes/MapReduce/MapReduceResultsStore.cs (hunk 2 above)
Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/MapReduce/MapReduceResultsStore.cs:66]: Okay, I understand. You're right it might affect the performance - I haven't thought about it. But still don't understand  how `_alreadyInitializedTrees.Contains(treeName)` can return `true` as `InitializeTree ` is called once

Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/MapReduce/MapReduceResultsStore.cs:66]: Edited the above comment: can return false -> can return true

Comment (redknightlois) [reply src/Raven.Server/Documents/Indexes/MapReduce/MapReduceResultsStore.cs:66]: Because now there is a single HashMap per each per `MapReduceStore`. Having said that will see if that condition ever triggers, I started trying to figure out where was the multiplicative factor be generated, and could be useless.

Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/MapReduce/MapReduceResultsStore.cs:66]: ok

Comment (redknightlois) [reply src/Raven.Server/Documents/Indexes/MapReduce/MapReduceResultsStore.cs]: Is not reusing pages at all a viable strategy?

File: src/Raven.Server/Documents/Indexes/MapReduce/MapReduceResultsStore.cs (hunk 1 above)
Comment (redknightlois) [reply src/Raven.Server/Documents/Indexes/MapReduce/MapReduceResultsStore.cs:64]: Is not reusing pages at all a viable strategy? What about having a shared FreedPages among `MapReduceResultsStore`?

Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/MapReduce/MapReduceResultsStore.cs:64]: Shared `FreedPages` makes sense to me. Since this code:

https://github.com/ravendb/ravendb/blob/436d63cd0b9b42473d7dc0945b8f4ebd2be6b2a3/src/Raven.Server/Documents/Indexes/MapReduce/ReduceMapResultsBase.cs#L332-L336

Doesn't need to be run in "contex" of any particular tree, we can run it before starting processing any tree (`HandleTreeReduction`) just once. That should be fine I think (tests will validate that for sure).

Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/MapReduce/MapReduceResultsStore.cs:64]: One potential problem, `HandleTreeReduction` might be called recursively

https://github.com/ravendb/ravendb/blob/436d63cd0b9b42473d7dc0945b8f4ebd2be6b2a3/src/Raven.Server/Documents/Indexes/MapReduce/ReduceMapResultsBase.cs#L406-L427

So maybe we'll be better of by calling `table.DeleteByKey(pageNumberSlice); ` for shared `FreedPages` _after_ processing all trees

**Comment by redknightlois** [comment]
@arekpalinski I implemented the shared free pages. It currently passes all fast tests and the indexing ones, but I am not entirely sure I am not missing something, because I haven't had to modify anything. Either the code is already handling the most complex case, so now it works as is OR I am missing something.

**Comment by arekpalinski** [comment]
I have created https://github.com/ravendb/ravendb/pull/8230 and applied some changes on top of yours. The most important one is moving `table.DeleteByKey(pageNumberSlice);` outside iterating over the reduce trees. Please review

**Comment by ayende** [comment]
I mreged https://github.com/ravendb/ravendb/pull/8230

Do we have something here that we still need?
